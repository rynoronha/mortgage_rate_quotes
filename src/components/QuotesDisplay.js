import React from 'react';
import { useSelector } from "react-redux";

function QuotesDisplay() {
    const quotes = useSelector(state => state.quotes);

    const formatPercent = (number) => {
        let decimalPlaces = number.toString().split('.')[1];
        
        if(!decimalPlaces || decimalPlaces.length <= 2) {
            return number.toFixed(2) + "%";
        }
        
        return parseFloat(number.toFixed(3)) + "%";
    }

    const formatCurrency = (number) => {
        return "$" + Math.round(number).toLocaleString('en-US');
    }

    return (
        <div className="table-container">
            <table className="table">
                    <thead>
                        <tr>
                        <th>LENDER</th>
                        <th>PRODUCT</th>
                        <th>RATE</th>
                        <th>CLOSING<br/>COSTS</th>
                        <th>MONTHLY<br/>PAYMENT</th>
                        <th>APR<br/></th>
                    </tr>
                    </thead>
                    <tbody>
                        {quotes.map((quote) => (
                            <tr key={quote.quote_id}>
                                <td>{quote.lenderName}</td>
                                <td>{quote.loanType}</td>
                                <td>{formatPercent(quote.interestRate)}</td>
                                <td>{formatCurrency(quote.closingCosts)}</td>
                                <td>{formatCurrency(quote.monthlyPayment)}</td>
                                <td>{formatPercent(quote.apr)}</td>
                            </tr>
                        ))}
                        
                    </tbody>
                </table>
                {!quotes.length &&
                    <div className="fill-out-form-notice">Please fill out the form above to display your quotes</div>
                }
        </div>
    )
}

export default QuotesDisplay
