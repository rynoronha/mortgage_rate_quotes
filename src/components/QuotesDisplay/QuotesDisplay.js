import React from 'react';
import { useSelector } from "react-redux";

function QuotesDisplay() {
    const quotes = useSelector(state => state.quotes);
    const isLoading = useSelector(state => state.isLoading)
    const areNoQuotesReturned = useSelector(state => state.areNoQuotesReturned)
    const error = useSelector(state => state.error)

    const formatPercent = (number) => {
        let decimalPlaces = number.toString().split('.')[1];
        
        if(!decimalPlaces || decimalPlaces.length <= 2) {
            return number.toFixed(2) + "%";
        }
        
        return parseFloat(number.toFixed(3)) + "%";
    }

    const formatCurrency = (number) => {
        return "$" + Math.round(number).toLocaleString('en-US');
    }

    return (
        <div className="table-container">
            <table data-testid="table" className="table">
                    <thead>
                        <tr>
                        <th>LENDER</th>
                        <th>PRODUCT</th>
                        <th>RATE</th>
                        <th>CLOSING<br/>COSTS</th>
                        <th>MONTHLY<br/>PAYMENT</th>
                        <th>APR<br/></th>
                    </tr>
                    </thead>
                    <tbody>
                        {!isLoading && !error && quotes.map((quote, i) => (
                            <tr key={i}>
                                <td>{quote.lenderName}</td>
                                <td>{quote.loanType}</td>
                                <td>{formatPercent(quote.interestRate)}</td>
                                <td>{formatCurrency(quote.closingCosts)}</td>
                                <td>{formatCurrency(quote.monthlyPayment)}</td>
                                <td>{formatPercent(quote.apr)}</td>
                            </tr>
                        ))}
                        
                    </tbody>
                </table>
                {!quotes.length && !isLoading && !error && !areNoQuotesReturned &&
                    <div data-testid="fill-out-form-notice" className="fill-out-form-notice">Please fill out the form above to display your quotes</div>
                }
                {isLoading &&
                    <div className="loading">Loading...</div>
                }
                {areNoQuotesReturned &&
                    <div className="no-quotes">Your search returned no results. <br/>
                        Please modify your query and try again.
                    </div>
                }
                {error &&
                    <div className="error-message">WHOOPS, something went wrong! Please try again. <br/>
                        <p className="error-status">STATUS CODE: {error.status}</p>
                    </div>
                }
        </div>
    )
}

export default QuotesDisplay
