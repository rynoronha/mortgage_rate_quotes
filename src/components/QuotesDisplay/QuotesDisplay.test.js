import React from 'react';
import  QuotesDisplay  from './QuotesDisplay';
import { render, screen, cleanup  } from '../../test-utils'

afterEach(cleanup);

test('renders the display table and fill out form notice', () => {
    const { queryByTestId } = render(<QuotesDisplay />);

    expect(queryByTestId("table")).toBeTruthy();
    expect(queryByTestId("fill-out-form-notice")).toBeTruthy();
});

test('it matches snapshot', () => {
    const tree = render(<QuotesDisplay />);

    expect(tree).toMatchSnapshot();
})

test('displays the results in the table when the quotes array is set', () => {
    render(<QuotesDisplay />, {initialState: { 
        quotes: [
            {
                lenderName: "lender1",
                loanType: "15YR FIXED",
                interestRate: 3.5,
                closingCosts: 5000,
                monthlyPayment: 2000,
                apr: 3.75
            },
            {
                lenderName: "lender2",
                loanType: "30YR FIXED",
                interestRate: 4.0,
                closingCosts: 6000,
                monthlyPayment: 3000,
                apr: 4.5
            }
        ]
    }});

    //1st row
    expect(screen.getByText(/lender1/)).toBeInTheDocument();
    expect(screen.getByText(/15YR FIXED/)).toBeInTheDocument();
    expect(screen.getByText(/3.5/)).toBeInTheDocument();
    expect(screen.getByText(/5,000/)).toBeInTheDocument();
    expect(screen.getByText(/2,000/)).toBeInTheDocument();
    expect(screen.getByText(/3.75/)).toBeInTheDocument();
    //2ndrow
    expect(screen.getByText(/lender2/)).toBeInTheDocument();
    expect(screen.getByText(/30YR FIXED/)).toBeInTheDocument();
    expect(screen.getByText(/4.0/)).toBeInTheDocument();
    expect(screen.getByText(/6,000/)).toBeInTheDocument();
    expect(screen.getByText(/3,000/)).toBeInTheDocument();
    expect(screen.getByText(/4.5/)).toBeInTheDocument();
})

test('displays properly formatted rates', () => {
    render(<QuotesDisplay />, {initialState: { 
        quotes: [
            {
                lenderName: "lender1",
                loanType: "15YR FIXED",
                interestRate: 3.5,
                closingCosts: 5000,
                monthlyPayment: 2000,
                apr: 3.75
            }
        ]
    }});

    expect(screen.getByText(/3.50%/)).toBeInTheDocument();
    expect(screen.getByText(/3.75%/)).toBeInTheDocument();
})

test('displays properly formatted currency values', () => {
    render(<QuotesDisplay />, {initialState: { 
        quotes: [
            {
                lenderName: "lender1",
                loanType: "15YR FIXED",
                interestRate: 3.5,
                closingCosts: 5000,
                monthlyPayment: 2000,
                apr: 3.75
            }
        ]
    }});

    expect(screen.getByText(/\$5,000/)).toBeInTheDocument();
    expect(screen.getByText(/\$2,000/)).toBeInTheDocument();
})

test('displays loading text when waiting for quotes to be fetched', () => {
    render(<QuotesDisplay />, {initialState: { 
        quotes: [],
        isLoading: true,
        areNoQuotesReturned: false,
        error: null
    }});

    expect(screen.getByText(/Loading.../)).toBeInTheDocument();
})

test('displays no results text when no quotes are returned', () => {
    render(<QuotesDisplay />, {initialState: { 
        quotes: [],
        isLoading: false,
        areNoQuotesReturned: true,
        error: null
    }});

    expect(screen.getByText(/Your search returned no results/)).toBeInTheDocument();
})

test('displays error text when there is an error fetching quotes', () => {
    render(<QuotesDisplay />, {initialState: { 
        quotes: [],
        isLoading: false,
        areNoQuotesReturned: false,
        error: {
            status: 403
        }
    }});

    expect(screen.getByText(/WHOOPS, something went wrong! Please try again/)).toBeInTheDocument();
})