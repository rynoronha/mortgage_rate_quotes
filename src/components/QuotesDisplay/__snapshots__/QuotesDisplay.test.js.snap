// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`it matches snapshot 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="table-container"
      >
        <table
          class="table"
          data-testid="table"
        >
          <thead>
            <tr>
              <th>
                LENDER
              </th>
              <th>
                PRODUCT
              </th>
              <th>
                RATE
              </th>
              <th>
                CLOSING
                <br />
                COSTS
              </th>
              <th>
                MONTHLY
                <br />
                PAYMENT
              </th>
              <th>
                APR
                <br />
              </th>
            </tr>
          </thead>
          <tbody />
        </table>
        <div
          class="fill-out-form-notice"
          data-testid="fill-out-form-notice"
        >
          Please fill out the form above to display your quotes
        </div>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="table-container"
    >
      <table
        class="table"
        data-testid="table"
      >
        <thead>
          <tr>
            <th>
              LENDER
            </th>
            <th>
              PRODUCT
            </th>
            <th>
              RATE
            </th>
            <th>
              CLOSING
              <br />
              COSTS
            </th>
            <th>
              MONTHLY
              <br />
              PAYMENT
            </th>
            <th>
              APR
              <br />
            </th>
          </tr>
        </thead>
        <tbody />
      </table>
      <div
        class="fill-out-form-notice"
        data-testid="fill-out-form-notice"
      >
        Please fill out the form above to display your quotes
      </div>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
